packages/driver/src/cy/commands/files.ts

```diff
@@ -4,118 +4,132 @@ import { basename } from 'path'
 import $errUtils from '../../cypress/error_utils'
 import type { Log } from '../../cypress/log'
 
-interface InternalReadFileOptions extends Partial<Cypress.Loggable & Cypress.Timeoutable> {
-  _log?: Log
-  encoding: Cypress.Encodings
+interface ReadFileOptions extends Partial<Cypress.Loggable & Cypress.Timeoutable> {
+  encoding?: Cypress.Encodings
 }
 
 interface InternalWriteFileOptions extends Partial<Cypress.WriteFileOptions & Cypress.Timeoutable> {
   _log?: Log
 }
 
 export default (Commands, Cypress, cy, state) => {
-  Commands.addAll({
-    readFile (file, encoding, userOptions: Partial<Cypress.Loggable & Cypress.Timeoutable> = {}) {
-      if (_.isObject(encoding)) {
-        userOptions = encoding
-        encoding = undefined
+  Commands.addQuery('readFile', function readFile (file, encoding, options: ReadFileOptions = {}) {
+    if (_.isObject(encoding)) {
+      options = encoding
+      encoding = options.encoding
+    }
+
+    encoding = encoding === undefined ? 'utf8' : encoding
+
+    const timeout = options.timeout ?? Cypress.config('defaultCommandTimeout')
+
+    this.set('timeout', timeout)
+    this.set('ensureExistenceFor', 'subject')
+
+    const log = options.log !== false && Cypress.log({ message: file, timeout })
+
+    if (!file || !_.isString(file)) {
+      $errUtils.throwErrByPath('files.invalid_argument', {
+        args: { cmd: 'readFile', file },
+      })
+    }
+
+    let fileResult: any = null
```

Comment:
The way this now works is basically a state machine, with three pieces of state:
1. The file read from disk
2. The currently active promise, waiting for a file from disk
3. The most recent error

The state can be updated in three different places.
1. When the query function executes (`return () => {`), if we have a `fileResult`, return it. Otherwise, create a new `filePromise` if we don't already have one (`createFilePromise()`) and throw `mostRecentError`.
2. When `filePromise` resolves (`.then((result) => {`), we set the contents to `fileResult`, and clear `filePromise`.
3. When `filePromise` rejects:
  a. If it rejects ((`.catch((err) => {`)) with "file doesn't exist", we set `fileResult` to "no file exists" and clear `filePromise`.
  b. Otherwise, we set the error to `mostRecentError`, and clear `filePromise`.
4. If an upcoming assertion fails, we clear `fileResult`.

The end result is that we query the disk for a file, and throw an error until we get back a result (including "no file exists"); if the server responds with an unexpected error, we retry.

If it succeeds, we start returning that file as the result.

If an upcoming assertion fails (including the implicit "file should exist" assertion), we clear the result and head back into the retry loop.